rosbag2_cpp/src/rosbag2_cpp/clocks/time_controller_clock.cpp

```diff
@@ -55,45 +55,72 @@ class TimeControllerClockImpl
     std::chrono::steady_clock::time_point steady;
   };
 
-  explicit TimeControllerClockImpl(PlayerClock::NowFunction now_fn)
-  : now_fn(now_fn)
+  explicit TimeControllerClockImpl(
+    PlayerClock::NowFunction now_fn, std::chrono::milliseconds sleep_time_while_paused)
+  : now_fn(now_fn),
+    sleep_time_while_paused(sleep_time_while_paused)
   {}
   virtual ~TimeControllerClockImpl() = default;
 
   template<typename T>
-  rcutils_duration_value_t duration_nanos(const T & duration)
+  rcutils_duration_value_t duration_nanos(const T & duration) const
   {
     return std::chrono::duration_cast<std::chrono::nanoseconds>(duration).count();
   }
 
-  rcutils_time_point_value_t steady_to_ros(std::chrono::steady_clock::time_point steady_time)
+  rcutils_time_point_value_t steady_to_ros(std::chrono::steady_clock::time_point steady_time) const
   RCPPUTILS_TSA_REQUIRES(state_mutex)
   {
     return reference.ros + static_cast<rcutils_duration_value_t>(
       rate * duration_nanos(steady_time - reference.steady));
   }
 
-  std::chrono::steady_clock::time_point ros_to_steady(rcutils_time_point_value_t ros_time)
+  std::chrono::steady_clock::time_point ros_to_steady(rcutils_time_point_value_t ros_time) const
   RCPPUTILS_TSA_REQUIRES(state_mutex)
   {
     const auto diff_nanos = static_cast<rcutils_duration_value_t>(
       (ros_time - reference.ros) / rate);
     return reference.steady + std::chrono::nanoseconds(diff_nanos);
   }
 
+  rcutils_time_point_value_t ros_now() const
+  RCPPUTILS_TSA_REQUIRES(state_mutex)
+  {
+    if (paused) {
+      return reference.ros;
+    }
+    return steady_to_ros(now_fn());
```

Comment:
Can `now_fn` ever be null?

Reply:
No - it's `const std::function` by value - assigned in the constructor.

Reply:
I looked into this for the case if null is passed in the constructor.
It looks like `std::function` cannot be assigned to null. It can be assigned to `nullptr`, where it then becomes "empty".
Calling an empty std::function will throw std::bad_function_call.

Reply:
Thanks for looking into it! Do you think it would be wise to add an `assert(now_fn != nullptr)` of some sort in the `TimeControllerClockImpl` constructor?

Reply:
Good call - opened https://github.com/ros2/rosbag2/pull/731 to handle this